util/src/main/java/com/google/cloud/hadoop/util/GrpcErrorTypeExtractor.java

```diff
@@ -34,8 +39,42 @@ public ErrorType getErrorType(Exception error) {
         return ErrorType.NOT_FOUND;
       case OUT_OF_RANGE:
         return ErrorType.OUT_OF_RANGE;
+      case ALREADY_EXISTS:
+        return ErrorType.ALREADY_EXISTS;
+      case FAILED_PRECONDITION:
+        return ErrorType.FAILED_PRECONDITION;
       default:
         return ErrorType.UNKNOWN;
     }
   }
+
+  @Override
+  public boolean bucketAlreadyExists(Exception e) {
+    ErrorType errorType = getErrorType(e);
+    if (errorType == ErrorType.ALREADY_EXISTS) {
+      return true;
+    }
+    // The gRPC API currently throws a FAILED_PRECONDITION status code instead of ALREADY_EXISTS,
+    // so we handle both these conditions in the interim.
+    // TODO: remove once the status codes are fixed.
+    else if (errorType == ErrorType.FAILED_PRECONDITION) {
+      StatusRuntimeException statusRuntimeException = getStatusRuntimeException(e);
+      return statusRuntimeException != null
+          && BUCKET_ALREADY_EXISTS_MESSAGE.equals(statusRuntimeException.getMessage());
+    }
+    return false;
+  }
+
+  /** Extracts StatusRuntimeException from the Exception, if it exists. */
+  @Nullable
+  private StatusRuntimeException getStatusRuntimeException(Exception e) {
+    Throwable cause = e;
+    do {
+      if (cause instanceof StatusRuntimeException) {
+        return (StatusRuntimeException) cause;
+      }
+      cause = cause.getCause();
+    } while (cause != null && cause != e);
```

Comment:
This will catch only self loops, correct? Why not have a counter of say 1000 and break after that.

Reply:
Done